File: app/models/fund.py
import uuid
from datetime import date, datetime, timezone
from enum import Enum as PyEnum

from sqlalchemy import (
    UUID, Date, DateTime, Enum, ForeignKey,
    Integer, Numeric, String
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class FundStatus(PyEnum):
    fundraising = "fundraising"
    investing = "investing"
    harvesting = "harvesting"
    closed = "closed"


class LPType(PyEnum):
    individual = "individual"
    institutional = "institutional"
    family_office = "family_office"
    fund_of_funds = "fund_of_funds"


class InvestmentType(PyEnum):
    lead = "lead"
    co_invest = "co_invest"
    follow_on = "follow_on"


class ScenarioType(PyEnum):
    base = "base"
    high = "high"
    downside = "downside"


class Fund(Base):
    __tablename__ = "funds"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    vintage_year: Mapped[int] = mapped_column(Integer, nullable=False)
    target_size_usd: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    deployed_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False, default=0)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    geography_focus: Mapped[str | None] = mapped_column(String(255), nullable=True)
    stage_focus: Mapped[str | None] = mapped_column(String(255), nullable=True)
    status: Mapped[FundStatus] = mapped_column(
        Enum(FundStatus, name="fund_status"), nullable=False
    )

    lps: Mapped[list["LP"]] = relationship(back_populates="fund")
    investments: Mapped[list["Investment"]] = relationship(back_populates="fund")
    scenarios: Mapped[list["FundScenario"]] = relationship(back_populates="fund")
    metrics: Mapped[list["FundMetric"]] = relationship(back_populates="fund")


class LP(Base):
    __tablename__ = "lps"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    fund_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("funds.id", ondelete="RESTRICT"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    type: Mapped[LPType] = mapped_column(
        Enum(LPType, name="lp_type"), nullable=False
    )
    committed_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)
    called_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False, default=0)
    distributed_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False, default=0)
    joined_at: Mapped[date] = mapped_column(Date, nullable=False)
    contact_email: Mapped[str | None] = mapped_column(String(255), nullable=True)

    fund: Mapped["Fund"] = relationship(back_populates="lps")


class Investment(Base):
    __tablename__ = "investments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    startup_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("startups.id", ondelete="RESTRICT"), nullable=False
    )
    fund_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("funds.id", ondelete="RESTRICT"), nullable=False
    )
    round_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("funding_rounds.id", ondelete="SET NULL"), nullable=True
    )
    amount_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)
    pre_money_val: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    equity_pct: Mapped[float] = mapped_column(Numeric(7, 4), nullable=False)
    date: Mapped[date] = mapped_column(Date, nullable=False)
    investment_type: Mapped[InvestmentType] = mapped_column(
        Enum(InvestmentType, name="investment_type"), nullable=False
    )
    follow_on_reserve_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False, default=0)

    startup: Mapped["Startup"] = relationship(foreign_keys=[startup_id])
    fund: Mapped["Fund"] = relationship(back_populates="investments")
    round: Mapped["FundingRound"] = relationship(foreign_keys=[round_id])


class FundScenario(Base):
    __tablename__ = "fund_scenarios"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    fund_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("funds.id", ondelete="CASCADE"), nullable=False
    )
    scenario_type: Mapped[ScenarioType] = mapped_column(
        Enum(ScenarioType, name="scenario_type"), nullable=False
    )
    moic_p25: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    moic_p50: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    moic_p75: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    irr_p25: Mapped[float] = mapped_column(Numeric(8, 6), nullable=False)
    irr_p50: Mapped[float] = mapped_column(Numeric(8, 6), nullable=False)
    irr_p75: Mapped[float] = mapped_column(Numeric(8, 6), nullable=False)
    dpi_projected: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    tvpi_projected: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    run_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc)
    )
    assumptions_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    fund: Mapped["Fund"] = relationship(back_populates="scenarios")


class FundMetric(Base):
    __tablename__ = "fund_metrics"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    fund_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("funds.id", ondelete="CASCADE"), nullable=False
    )
    calculation_date: Mapped[date] = mapped_column(Date, nullable=False)
    moic: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    tvpi: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    dpi: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    rvpi: Mapped[float] = mapped_column(Numeric(10, 4), nullable=False)
    irr: Mapped[float | None] = mapped_column(Numeric(8, 6), nullable=True)
    total_invested_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)
    total_fmv_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)
    total_realized_usd: Mapped[float] = mapped_column(Numeric(18, 2), nullable=False)

    fund: Mapped["Fund"] = relationship(back_populates="metrics")

File: frontend/lib/mock/fund.ts
import type { FundData, Investment, FundMetrics, ScenarioResult } from '../types'

export const fund: FundData = {
  id: 'fund-0001-0001-0001-000000000001',
  name: 'AIDA Ventures Fund I',
  vintage_year: 2023,
  target_size_usd: 10000000,
  deployed_usd: 3150000,
  status: 'investing',
}

export const investments: Investment[] = [
  { id: 'inv-001', startup_name: 'FinStack',  amount_usd: 300000,  stage_at_entry: 'Seed',     entry_date: '2023-08-15', ownership_pct: 8.5  },
  { id: 'inv-002', startup_name: 'LogiFlow',  amount_usd: 750000,  stage_at_entry: 'Series A', entry_date: '2023-11-20', ownership_pct: 5.2  },
  { id: 'inv-003', startup_name: 'MediSync',  amount_usd: 150000,  stage_at_entry: 'Pre-Seed', entry_date: '2024-01-10', ownership_pct: 12.0 },
  { id: 'inv-004', startup_name: 'CreditIA',  amount_usd: 400000,  stage_at_entry: 'Seed',     entry_date: '2024-02-28', ownership_pct: 9.0  },
  { id: 'inv-005', startup_name: 'AgriSense', amount_usd: 250000,  stage_at_entry: 'Seed',     entry_date: '2024-04-05', ownership_pct: 7.5  },
  { id: 'inv-006', startup_name: 'EduStack',  amount_usd: 200000,  stage_at_entry: 'Pre-Seed', entry_date: '2024-05-12', ownership_pct: 15.0 },
  { id: 'inv-007', startup_name: 'LegalBot',  amount_usd: 100000,  stage_at_entry: 'Pre-Seed', entry_date: '2024-07-01', ownership_pct: 10.0 },
  { id: 'inv-008', startup_name: 'FleetOS',   amount_usd: 350000,  stage_at_entry: 'Seed',     entry_date: '2024-08-20', ownership_pct: 8.0  },
  { id: 'inv-009', startup_name: 'TaxFlow',   amount_usd: 200000,  stage_at_entry: 'Pre-Seed', entry_date: '2024-10-15', ownership_pct: 11.0 },
]

export const metrics: FundMetrics = {
  moic_current: 2.83,
  irr_estimate_pct: 22.4,
  total_invested_usd: 3150000,
  portfolio_count: 10,
}

export const savedScenarios: ScenarioResult[] = [
  {
    label: 'conservador',
    moic_p25: 1.1,
    moic_p50: 1.8,
    moic_p75: 2.4,
    irr_p25: 6.2,
    irr_p50: 14.8,
    irr_p75: 21.3,
    total_invested_usd: 3150000,
  },
  {
    label: 'base',
    moic_p25: 1.6,
    moic_p50: 2.83,
    moic_p75: 4.1,
    irr_p25: 12.4,
    irr_p50: 22.4,
    irr_p75: 34.1,
    total_invested_usd: 3150000,
  },
  {
    label: 'optimista',
    moic_p25: 2.8,
    moic_p50: 4.2,
    moic_p75: 6.5,
    irr_p25: 24.1,
    irr_p50: 35.2,
    irr_p75: 52.8,
    total_invested_usd: 3150000,
  },
]
